### ISSUE TITLE
Do iterators already provide everything we need?

### ISSUE THREAD

**Issue by jnicklas** [thefrontside/effection#92]
We were talking yesterday about this idea of using something like a continuation to represent the different stages that an execution context goes through. I spent some time thinking about this today, and I started to think that maybe this is all unnecessary. Do iterators already provide us with everything we need?

Let's walk through an example of an iterator:

``` javascript
gen = function*() {
  try {
    yield 1;
    yield 2;
  } catch(e) {
    yield "error";
    throw e;
  } finally {
    yield "finally";
  }
}
```

This is the successful case:

``` javascript
> a = gen()
Object [Generator] {}
> a.next()
{ value: 1, done: false }
> a.next()
{ value: 2, done: false }
> a.next()
{ value: 'finally', done: false }
> a.next()
{ value: undefined, done: true }
```

And let's step through how this behaves in case of throwing an error:

``` javascript
> a = gen()
Object [Generator] {}
> a.next()
{ value: 1, done: false }
> a.next()
{ value: 2, done: false }
> a.throw(new Error("foo"))
{ value: 'error', done: false }
> a.next()
{ value: 'finally', done: false }
> a.next()
Thrown:
Error: foo
> a.next()
{ value: undefined, done: true }
```

And let's step through how this behaves in case of return:

```
> a = gen()
Object [Generator] {}
> a.next()
{ value: 1, done: false }
> a.next()
{ value: 2, done: false }
> a.return()
{ value: 'finally', done: false }
> a.next()
{ value: undefined, done: true }
```

In other words, iterators/generators already provide us with a way of doing cleanup and error handling.

What if `halt` and `fail` didn't just return/throw into those iterators, but instead also proceeded to continue driving those generators to conclusion by continuing to call `next`? This coupled with a timeout mechanism which would terminate after a certain amount of time would essentially give us async cleanup for free.

Now we only need some mechanism to invert a generator, so we can move from callbacks to generators, and suddenly we don't actually need anything else.

@cowboyd I'm interested in hearing your thoughts on this!

**Comment by cowboyd** [comment]
I think the main thing we need is a way to stitch generators together into a coherent stack. I tried to write down some of these explorations here https://github.com/cowboyd/computation

**Comment by jnicklas** [comment]
This is precisely the approach we've chosen for v2.
